File: 0028_Implement_strStr()/main.cpp
// KMP
class Solution {
  public:
    vector<int> next;
    int strStr(string haystack, string needle) {
        if (needle.empty()) {
            return 0;
        }
        int n = haystack.size(), m = needle.size();
        string s = " " + haystack, p = " " + needle;
        next.resize(m + 1);

        for (int i = 2, j = 0; i <= m; i++) {
            while (j != 0 && p[i] != p[j + 1]) {
                j = next[j];
            }
            if (p[i] == p[j + 1]) {
                j++;
            }
            next[i] = j;
        }

        for (int i = 1, j = 0; i <= n; i++) {
            while (j != 0 && s[i] != p[j + 1]) {
                j = next[j];
            }
            if (s[i] == p[j + 1]) {
                j++;
            }
            if (j == m) {
                return i - m;
            }
        }
        return -1;
    }
};

File: 0028_Implement_strStr()/golang/main.go
func strStr(haystack string, needle string) int {
	if len(haystack) == 0 && len(needle) == 0 {
		return 0
	}
	n, m := len(haystack), len(needle)
	for i := 0; i <= n-m; i++ {
		j, k := i, 0
		for j < n && k < m && haystack[j] == needle[k] {
			j++
			k++
		}
		if k == m {
			return i
		}
	}
	return -1
}